docs/tutorials/typescript/workqueue-slackbot/code/temporal-application/src/workflows/workqueue.ts

```diff
@@ -0,0 +1,67 @@
+import {
+  CancellationScope,
+  CancelledFailure,
+  continueAsNew,
+  defineQuery,
+  defineSignal,
+  setHandler,
+  workflowInfo,
+  condition,
+} from "@temporalio/workflow";
+import {WorkqueueData} from "../../../common-types/types";
+
+const getWorkqueueDataQuery = defineQuery<WorkqueueData[]>("getWorkqueueData");
+const addWorkqueueDataSignal =
+  defineSignal<[WorkqueueData]>("addWorkqueueData");
+const claimWorkSignal =
+  defineSignal<[{workId: string; claimantId: string}]>("claimWork");
+const completeWorkSignal = defineSignal<[{workId: string}]>("completeWork");
+
+export async function workqueue(existingData?: WorkqueueData[]): Promise<void> {
+  const wqdata: WorkqueueData[] = existingData ?? [];
+
+  // Register a Query handler for 'getWorkqueueData'
+  setHandler(getWorkqueueDataQuery, () => {
+    return wqdata;
+  });
+
+  // Register the Signal handler for adding work
+  setHandler(addWorkqueueDataSignal, (data: WorkqueueData) => {
+    wqdata.push(data);
+  });
+
+  // Register Signal handler for claiming work
+  setHandler(claimWorkSignal, ({workId, claimantId}) => {
+    const workItem = wqdata.find((item) => item.id === workId);
+    if (workItem) {
+      workItem.claimantId = claimantId;
+      workItem.status = 2;
+    }
+  });
+
+  // Register Signal handler for completing work
+  setHandler(completeWorkSignal, ({workId}) => {
+    const index = wqdata.findIndex((item) => item.id === workId);
+    if (index !== -1) {
+      wqdata.splice(index, 1);
+    }
+  });
+
+  while (!workflowInfo().continueAsNewSuggested) {
+    try {
+      while (true) {
+        // Await cancellation
+        await CancellationScope.current().cancelRequested;
```

Comment:
The `while (true)` loop around this is completely useless at the moment. `await CancellationScope.current().cancelRequested` can only throw.

Also, the `!workflowInfo().continueAsNewSuggested` above will never get rechecked since there is no way for the content of the contained block to complete and loop.

Reply:
Instead of all of this (the `while (!workflowInfo().continueAsNewSuggested) { ... }`), I think you should be able to do:

```
  // Await until the workflow gets cancelled, history grow too much
  await Promise.any([
    CancellationScope.current().cancelRequested,
    condition(() => workflowInfo().continueAsNewSuggested),
  ])
```

Reply:
Actually, `condition` is already cancellation-aware... So you can simply do this:

```
try {
  await condition(() => workflowInfo().continueAsNewSuggested);
} catch (e) {
  // This is just for example, as you end up rethrowing the exception in both cases
  if (isCancellation(e)) { ... }
  else { ... }
}
```

Reply:
this was super helpful